src/chat/chatAgentRequestHandler.ts

```diff
@@ -0,0 +1,633 @@
+                        );
+                        console.warn(`Unhandled message type: ${result.messageType}`);
+                        continuePollingMessages = false;
+                        break;
+                }
+
+                Utils.logDebug(`Done processing message for '${conversationUri}'`);
+                // Output reply text if needed
+                if (printTextout) {
+                    stream.markdown(replyText);
+                    printTextout = false;
+                }
+            }
+        } catch (err) {
+            handleError(err, stream, correlationId);
+        }
+
+        return { metadata: { command: "", correlationId: correlationId } };
+    };
+
+    async function processToolCalls(
+        sqlTools: { tool: LanguageModelChatTool; parameters: string }[],
+        conversationUri: string,
+        replyText: string,
+        copilotService: CopilotService,
+    ): Promise<GetNextMessageResponse> {
+        if (sqlTools.length === 0) {
+            throw new Error("No tools to process.");
+        }
+
+        let result: GetNextMessageResponse;
+        for (const toolCall of sqlTools) {
+            result = await copilotService.getNextMessage(
+                conversationUri,
+                replyText,
+                toolCall.tool,
+                toolCall.parameters,
+            );
+
+            if (result.messageType === MessageType.Complete) {
+                break;
+            }
+        }
+
+        return result!;
+    }
+
+    function prepareRequestMessages(
+        result: GetNextMessageResponse,
+        context: vscode.ChatContext,
+        referenceTexts: string[],
+    ): vscode.LanguageModelChatMessage[] {
+        // Separate system messages from the requestMessages
+        const systemMessages = result.requestMessages
+            .filter((message: LanguageModelRequestMessage) => message.role === MessageRole.System)
+            .map((message: LanguageModelRequestMessage) =>
+                vscode.LanguageModelChatMessage.Assistant(message.text),
+            );
+
+        const historyMessages = context.history
+            .map((historyItem) => {
+                if ("prompt" in historyItem) {
+                    return vscode.LanguageModelChatMessage.User(historyItem.prompt);
+                } else {
+                    const responseContent = historyItem.response
+                        .map((part) => ("content" in part ? part.content : ""))
+                        .join("");
+                    return responseContent.trim()
+                        ? vscode.LanguageModelChatMessage.Assistant(responseContent)
+                        : undefined;
+                }
+            })
+            .filter((msg): msg is vscode.LanguageModelChatMessage => msg !== undefined);
+
+        // Include the reference messages
+        // TODO: should we cut off the reference message or send a warning if it is too long? (especially without selection)
```

Comment:
What's the current limit/how is that determined?  At the very least, log the truncation.

What action could the user take if they see a warning about it being too long if it isn't a selection?
AI summary of the reference message would probably be best, but likely a pain to wire up.

Reply:
this is left as a TODO for now, so those questions haven't been resolved at this point